{"cells":[{"cell_type":"markdown","source":["# Delta Lakeによる大規模ゲノミクスパイプラインの簡略化\n\n#![Unified Analytics Platform for Genomics Architecture](https://pages.databricks.com/rs/094-YMS-629/images/hls_UAP4Genomics_architecture_WB.png)\n\nこのノートブックでは、Databricksにおける**大規模**ゲノミクスパイプラインのオペレーションと本格運用をどのように簡素化するのかを説明します。\n\n## シンプルな疑問に応えることすら困難です!\n\nゲノムデータはサイズが非常に大きいため、従来のシングルノードのバイオインフォマティクスツールでは以下のようなシンプルな疑問に答えるにも非常に時間がかかってしまいます。\n\n- どれだけのサンプルをシーケンシングしたのか？\n  - 今週は?\n  - 今月は?\n- それぞれいくつの変異体があったのか？\n- コホートにおいて何のクラスの変異体を観測したのか？\n\n## 大規模ゲノミクスパイプラインのデモンストレーション\n\nここではDelta Lakeにシーケンスデータのストリーミングを流し込み、リアルタイムで分析を行う様子をデモします。\n\nこのノートブックはi3.2xlargeインスタンスのドライバー、8台のワーカーで構成されるDatabricks 4.3(Apache Spark 2.3.1, Scala 2.11)クラスターでテストされています。\n\n**参考情報**\n- [Simplifying Genomics Pipelines at Scale with Databricks Delta \\- The Databricks Blog](https://databricks.com/blog/2019/03/07/simplifying-genomics-pipelines-at-scale-with-databricks-delta.html)\n- [Databricks Delta Lakeによる大規模ゲノミクスパイプラインの簡略化 \\- Qiita](https://qiita.com/taka_yayoi/items/2ae740ab884c26e5906e)\n\n<table>\n  <tr><th>作成者</th><th>Databricks Japan</th></tr>\n  <tr><td>作成日</td><td>2021/06/21</td></tr>\n  <tr><td>バージョン</td><td>1.0</td></tr>\n  <tr><td>DBR</td><td>8.3</td></tr>\n</table>\n\n<img style=\"margin-top:25px;\" src=\"https://jixjiadatabricks.blob.core.windows.net/images/databricks-logo-small-new.png\" width=\"140\">"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"dfaca604-44ab-4346-b5a1-0370cd77ca36"}}},{"cell_type":"code","source":["# ライブラリのインポート\nimport pyspark.sql.functions as fx\nfrom pyspark.sql.types import StringType, IntegerType, ArrayType\nimport numpy as np\nimport pandas as pd\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n# 設定\nsns.set(style=\"white\")\nspark.conf.set(\"spark.databricks.delta.preview.enabled\", \"true\")\nspark.conf.set(\"spark.databricks.delta.formatCheck.enabled\", \"false\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c44dba48-4aca-4c16-9d3b-30bf70c34e82"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["#\n# パース用関数\n#\n\n# 変異体のコーディング\ndef get_coding_mutations(df):\n  df = df.where(fx.col(\"proteinHgvs\").rlike(\"^[p\\.]\")). \\\n          where((fx.col(\"mutationType\") == \"nonsynonymous\") | \n                (fx.col(\"mutationType\") == \"synonymous\") | \n                (fx.col(\"effect\") == \"stop_gained\"))\n  return df\n\n# proteinHgvsのパース\ndef parse_proteinHgvs(hgvs):\n  \"\"\"\n  parses proteinHgvs string into amino acid substitutions\n  :param hgvs: str, proteinHgvs p.[codon1][position][codon2]\n  :return: list with two amino acids (* if stop codon)\n  \"\"\"\n  hgvs_list = list(hgvs)\n  aa1 = \"\".join(hgvs_list[2:5])\n  if (hgvs_list[-1] == \"*\"): # * = 終止コドン(stop codon): タンパク質合成を終了させるmRNA上のコドン(mRNA上のヌクレオチド3個から成る配列の単位)\n    aa2 = hgvs_list.pop(-1)\n  else:\n    aa2 = \"\".join(hgvs_list[-3:]) # 他のコドンは3文字の短縮形\n  return [aa1, aa2]\n\n# parse_proteinHgvsをUDF(ユーザー定義関数)として定義\nparse_proteinHgvs_udf = fx.udf(parse_proteinHgvs, ArrayType(StringType()))\n\n# アミノ酸交換の取得\ndef get_amino_acid_substitutions(df, hgvs_col):\n  \"\"\"\n  parse hgvs notation to get amino acid substitutions in a manageable format\n  \"\"\"\n  df = df.withColumn(\"tmp\", parse_proteinHgvs_udf(fx.col(hgvs_col))). \\\n          withColumn(\"reference\", fx.col(\"tmp\")[0]). \\\n          withColumn(\"alternate\", fx.col(\"tmp\")[1]). \\\n          drop(\"tmp\", hgvs_col)\n  return df\n\n# アミノ酸交換のカウント\ndef count_amino_acid_substitution_combinations(df):\n  df = df.groupBy(\"reference\", \"alternate\").count(). \\\n          withColumnRenamed(\"count\", \"substitutions\")\n  return df"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"functions","showTitle":false,"inputWidgets":{},"nuid":"e8e6b57e-720c-4e57-a7f3-f74eb7beb918"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## 初期セットアップ\n\nパイプラインを構築するにあたり、まずは以下の手順で単一の`sampleId`に対応する試験のParquetファイルをDelta Lakeテーブルに書き込みます。\n\n- Deltaストリームのパス`delta_stream_outpath`を指定してテーブルを作成します。\n- 単一の`sampleId`の試験のファイルを読み込み、`delta_stream_outpath`に書き込むSparkジョブを作成します。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"006eabdb-1118-47e8-9295-bb70b29de8a3"}}},{"cell_type":"code","source":["import re\nfrom pyspark.sql.types import * \n\n# ログインIDからUsernameを取得\nusername_raw = dbutils.notebook.entry_point.getDbutils().notebook().getContext().tags().apply('user')\n# Username の英数字以外を除去し、全て小文字化。Username をファイルパスやデータベース名の一部で使用可能にするため\nusername = re.sub('[^A-Za-z0-9]+', '', username_raw).lower()\n\n# データベース名\ndb_name = f\"{username}_genomics\"\n\n# Hiveメタストアのデータベースの準備:データベースの作成\nspark.sql(f\"CREATE DATABASE IF NOT EXISTS {db_name}\")\n# Hiveメタストアのデータベースの選択\nspark.sql(f\"USE {db_name}\")\n\n# ファイル格納パス\nwork_path = f\"dbfs:/tmp/{username}/dnaseq/\"\n\nprint(\"database name: \" + db_name)\nprint(\"path: \" + work_path)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"52c14869-af74-4ccb-9633-576551904cc4"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# Deltaストリームの出力パス\ndelta_stream_outpath = f\"{work_path}annotations_etl_delta_stream/\"\n\n# ソースとなるannotations_etl_parquet\nannotations_etl_parquet_path = \"/databricks-datasets/genomics/annotations_etl_parquet/\"\n\n# 単一のsampleIdを指定\nsingle_sampleId_path = annotations_etl_parquet_path + \"sampleId=SRS000030_SRR709972\"\n\n# ディレクトリがある場合には削除して再作成\ndbutils.fs.rm(delta_stream_outpath, True)\ndbutils.fs.mkdirs(delta_stream_outpath)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5bc07c6b-6b15-4770-8bd2-0d0167f17ce0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# 単一のsampleIdの読み込み\nspark.read.format(\"parquet\").load(single_sampleId_path).write.format(\"delta\").save(delta_stream_outpath)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"928be640-1060-4429-94f0-2cd3b81b3242"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## Delta Lakeでストリーミングを始める\n\n**重要!**\n- 以降のセルを実行するにはストリームの初期化が完了するのを待ってください。\n- 最初は単一サンプルの情報を見ることになります。\n- 後ほどDeltaテーブルにさらにサンプルを追加することで、テーブル、グラフがリアルタイムでアップデートされます。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"184e2159-80f5-4b50-8b73-3d7d80f3eb73"}}},{"cell_type":"markdown","source":["## シンプルな疑問に回答していきます\n\nDeltaストリームに問いかける質問には以下のものが含まれます：\n\n- 変異体のカウント\n- Deltaテーブルの最初の1000行の表示\n- テーブルのスキーマの確認"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"77436f04-e355-4843-9d0c-7ef59955684e"}}},{"cell_type":"code","source":["# 変異体のカウント\nexomes = spark.readStream.format(\"delta\").load(delta_stream_outpath)\ndisplay(exomes.groupBy(\"sampleId\").count().withColumnRenamed(\"count\", \"variants\"))"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"b97f8fe6-82b2-415f-babd-89e0a142f95a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# Deltaテーブル`exomes`の最初の1000行を表示\ndisplay(exomes)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c99a5b94-2e57-4f9b-b760-d05fc9c9c8ba"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# Delltaテーブル`exomes`のスキーマを確認\nexomes.printSchema()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"0864409a-7ea7-44b2-a67e-06a95a959498"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## 一塩基ヌクレオチド変異体のカウント\n\n一塩基ヌクレオチド変異体(SNV)のカウントを計算するために、以下のことを行います:\n- referenceAlleleとalternateAlleleのフィルタリング条件に基づいて`snvs`ストリーミングデータフレームを作成します。\n- 棒グラフでデータを参照するためにSpark SQLクエリーを実行します。\n  - これはストリーミングDeltaテーブルであり、`exomes`Deltaテーブルに新たなデータが追加されると棒グラフは継続的に更新されることに注意してください。\n  \n**参考情報**\n- [一塩基バリアント：日経バイオテクONLINE](https://bio.nikkeibp.co.jp/atcl/report/16/011900001/16/05/02/00022/)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"dd4b88da-5610-43e2-adcf-100ef5319a1c"}}},{"cell_type":"code","source":["snvs = exomes.where((fx.length(fx.col(\"referenceAllele\")) == 1) & \n                    (fx.length(fx.col(\"alternateAllele\")) == 1))\nsnvs.createOrReplaceTempView(\"snvs\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5915a64d-269e-4ad8-92db-b69a962fea4a"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["%sql\nselect referenceAllele, alternateAllele, count(1) as GroupCount \n  from snvs\n group by referenceAllele, alternateAllele\n order by GroupCount desc"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ed31a3fa-15d0-4802-9650-58002626ca44"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## 変異体のカウントは?\n\n変異体タイプのカウントを行うためには、Deltaテーブル`exomes`に対して`GROUP BY`を実行します。ここではドーナツチャートを使用します。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f33c1467-b9ab-47db-b332-e7b8a89818bf"}}},{"cell_type":"code","source":["display(exomes.groupBy(\"mutationType\").count())"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"76a5b233-6ab7-45d8-87a5-db97496c9e9b"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["**注意:** ご自身のユースケースでリアルタイムでの更新が必要ない場合には、シーケンシングを行う日次、週次の定期ジョブを実行することもできます。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"54f821b4-38c5-4808-885f-7e4dd665c3a7"}}},{"cell_type":"markdown","source":["## アミノ酸置換ヒートマップの作成\n\nアミノ酸置換ヒートマップを作成するために、*pandas*と*matplotlib*を使用します。\n\n**重要!**\n- 以降のセルを実行するにはストリームの初期化が完了するのを待ってください。\n- ビルトインの`display()`以外のプロット処理はストリーミングテーブルに対して動作しません。\n  - このため、一旦ストリームをメモリー(あるいはParquetに)に書き出して、手動で以下のセルを実行しアミノ酸置換ヒートマップを作成します。\n  \n**参考資料**\n- [アミノ酸置換とは何？ Weblio辞書](https://www.weblio.jp/content/%E3%82%A2%E3%83%9F%E3%83%8E%E9%85%B8%E7%BD%AE%E6%8F%9B#:~:text=%E3%83%8E%E3%83%B3%E3%82%B7%E3%83%8E%E3%83%8B%E3%83%9E%E3%82%B9%E3%81%AA%E3%82%B3%E3%83%BC%E3%83%89%E9%85%8D%E5%88%97%E4%B8%8A,%E3%81%AE%E3%82%A2%E3%83%9F%E3%83%8E%E9%85%B8%E3%81%AB%E5%A4%89%E3%82%8F%E3%82%8B%E3%81%93%E3%81%A8%E3%80%82)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"bd0c5629-1d0b-4622-b560-064c8bcd7417"}}},{"cell_type":"code","source":["# アミノ酸置換のストリームの構築\ncoding = get_coding_mutations(exomes)\naa_substitutions = get_amino_acid_substitutions(coding.select(\"proteinHgvs\"), \"proteinHgvs\")\naa_counts = count_amino_acid_substitution_combinations(aa_substitutions)\naa_counts. \\\n  writeStream. \\\n  format(\"memory\"). \\\n  queryName(\"amino_acid_substitutions\"). \\\n  outputMode(\"complete\"). \\\n  trigger(processingTime='60 seconds'). \\\n  start()"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5bda3aaf-52e5-4c35-b270-0dbfa77684c0"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# アミノ酸置換ヒートマップの作成\n# このヒートマップはpandasとmatplotlibを使用しているのでストリーミングテーブルでは動作しません\namino_acid_substitutions = spark.read.table(\"amino_acid_substitutions\")\nmax_count = amino_acid_substitutions.agg(fx.max(\"substitutions\")).collect()[0][0]\naa_counts_pd = amino_acid_substitutions.toPandas()\naa_counts_pd = pd.pivot_table(aa_counts_pd, values='substitutions', index=['reference'], columns=['alternate'], fill_value=0)\n\nfig, ax = plt.subplots()\nwith sns.axes_style(\"white\"):\n  ax = sns.heatmap(aa_counts_pd, vmax=max_count*0.4, cbar=False, annot=True, annot_kws={\"size\": 7}, fmt=\"d\")\nplt.tight_layout()\ndisplay(fig)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1e9dd3b3-e563-4b57-8437-1e7f7b0d0f8f"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## 時間経過における変異体の数の集計\n\n**重要!**\n- 以降のセルを実行するにはストリームの初期化が完了するのを待ってください。\n- ビルトインの`display()`以外のプロット処理はストリーミングテーブルに対して動作しません。\n  - このため、一旦ストリームをメモリー(あるいはParquetに)に書き出して、手動で以下のセルを実行しアミノ酸置換ヒートマップを作成します。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5bb146e8-139d-47b6-9618-98a17dff0952"}}},{"cell_type":"code","source":["# 参照: https://docs.databricks.com/spark/latest/structured-streaming/examples.html#foreachbatch-sqldw-example\n\n# パスの設定\nvariant_count_test_parquet_path = f\"{work_path}variant_count_test_parquet\"\n\n# Deltaフォーマットで書き出し\ndef write_to_delta(df, epochId):\n  df.write. \\\n     format(\"delta\"). \\\n     mode('append'). \\\n     save(variant_count_test_parquet_path)\n\n# 設定\nspark.conf.set(\"spark.sql.shuffle.partitions\", \"1\")\n\n# 60秒ごとにストリームを書き出すクエリーの実行\nquery = (\n  exomes.withColumn(\"mutations\", fx.lit(\"mutations\")). \\\n         select(\"contigName\", \"start\", \"referenceAllele\", \"alternateAllele\", \"mutations\"). \\\n         dropDuplicates(). \\\n         groupBy(\"mutations\"). \\\n         count(). \\\n         withColumn('time', fx.lit(fx.current_timestamp())). \\\n         writeStream. \\\n         trigger(processingTime='60 seconds'). \\\n         foreachBatch(write_to_delta). \\\n         outputMode(\"update\"). \\\n         start()\n    )"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"da98c650-45ab-46d1-92ea-b846269ad1e1"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["# 時間経過における別個の変異体の数\nvariants = spark.readStream.format(\"delta\").load(variant_count_test_parquet_path).withColumnRenamed(\"count\", \"distinct mutations\")\ndisplay(variants)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"e2d5f6a2-3ff6-4ea3-815b-2e6de45ec7b2"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## ゲノミクスパイプラインにデータを追加\n\nゲノミクスパイプラインの設定が終わったので、Deltaテーブルにさらにサンプルを追加します。\n- ストリーミングをセットアップしているので、上のグラフ(ヒートマップは除く)とテーブルはデータが投入される度に更新されます。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"a0b3f68f-4509-4c5a-9bff-d99fd6c497cc"}}},{"cell_type":"code","source":["import time\nfiles = dbutils.fs.ls(annotations_etl_parquet_path)\ncounter=0\n\n# 時間間隔を持って、すべてのParquetファイルに対してループします\ntime.sleep(10)\nfor sample in files:\n  counter+=1\n  annotation_path = sample.path\n  if (\"sampleId=\" in annotation_path):\n    sampleId = annotation_path.split(\"/\")[4].split(\"=\")[1]\n    variants = spark.read.format(\"parquet\").load(str(annotation_path))\n    print(\"running \" + sampleId)\n    if(sampleId != \"SRS000030_SRR709972\"):\n      variants.write.format(\"delta\"). \\\n             mode(\"append\"). \\\n             save(delta_stream_outpath)\n    time.sleep(10)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"ffba66da-8697-45a1-a23c-16c98e228788"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["## クリーンアップ\n終了したら生成したデータを削除しましょう。"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"1bd1645d-b9d0-4f51-bd35-eb3004e1acc8"}}},{"cell_type":"code","source":["%python\n\n# タイムアウト時間(millseconds)を変更\nspark.conf.set(\"spark.sql.streaming.stopTimeout\", 60000)\n\n# Sreaming処理を停止\nfor s in spark.streams.active:\n  print(\"Stream name:\", s.name)\n  s.stop()\n\nprint(\"Stopped.\")"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"c24f04f8-28d2-457f-a65e-b03720dcede9"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"code","source":["dbutils.fs.rm(delta_stream_outpath, True)\ndbutils.fs.rm(variant_count_test_parquet_path, True)"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"5de530f4-c3e7-43b5-88a2-86ec7021df76"}},"outputs":[{"output_type":"display_data","metadata":{"application/vnd.databricks.v1+output":{"data":"","errorSummary":"","metadata":{},"type":"ipynbError","arguments":{}}},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>"]}}],"execution_count":0},{"cell_type":"markdown","source":["# END"],"metadata":{"application/vnd.databricks.v1+cell":{"title":"","showTitle":false,"inputWidgets":{},"nuid":"f24378f6-1fae-4be3-99f2-9644d604ba2e"}}}],"metadata":{"application/vnd.databricks.v1+notebook":{"notebookName":"Delta Lakeによる大規模ゲノミクスパイプラインの簡略化","dashboards":[{"elements":[{"elementNUID":"dfaca604-44ab-4346-b5a1-0370cd77ca36","guid":"25364414-d316-4c3f-a960-3ff5e0e4c1ff","options":null,"position":{"x":0,"y":0,"height":9,"width":24,"z":null},"elementType":"command"},{"elementNUID":"b97f8fe6-82b2-415f-babd-89e0a142f95a","guid":"5cefb12d-41bb-4e0e-b066-5460bcb8ecb2","options":{"showTitle":true,"scale":0,"autoScaleImg":false,"title":"Variant Counts","titleAlign":"center"},"position":{"x":15,"y":9,"height":21,"width":9,"z":null},"elementType":"command"},{"elementNUID":"ed31a3fa-15d0-4802-9650-58002626ca44","guid":"8cceaa7f-b8b4-437e-a389-4d197ef5b840","options":{"showTitle":true,"scale":0,"autoScaleImg":false,"title":"Single Nucleotide Variant Count","titleAlign":"center"},"position":{"x":0,"y":9,"height":10,"width":15,"z":null},"elementType":"command"},{"elementNUID":"76a5b233-6ab7-45d8-87a5-db97496c9e9b","guid":"a23ccd17-2555-4b08-9a62-99933222f09a","options":{"showTitle":true,"scale":0,"autoScaleImg":false,"title":"Count of Mutation Types","titleAlign":"center"},"position":{"x":0,"y":19,"height":11,"width":15,"z":null},"elementType":"command"}],"guid":"03727bc0-7089-49ba-ac2b-e15910e15c1d","layoutOption":{"stack":true,"grid":true},"version":"DashboardViewV1","nuid":"10052a9c-2e49-463a-a56c-398644d7a18b","origId":2917837587014504,"title":"Overview","width":1024,"globalVars":{}}],"notebookMetadata":{"pythonIndentUnit":2},"language":"python","widgets":{},"notebookOrigID":2917837587014474}},"nbformat":4,"nbformat_minor":0}
